### PR TITLE
Improve Module Lesson

### PR DESCRIPTION
**Pull request by ybur-yug**
This includes:
- Unify style (previously was mixing `  ` and ` ` for spacing + not
  linebreaking per sentence as many other lessons have been in latest
  PRs, and use backticks on some words that were missed
- Mention `defoverridable` in `use` section
- Add example of overriding `use`'d function
- Mention `IEx.pry` as a common use case for `require`
- Mention lexical scope of `import`

## TODO
- [ ] Create translation issue once reviewed/merged

### CODE REVIEW

File: lessons/basics/modules.md [hunk 1]
```diff
@@ -194,9 +202,28 @@ import List, only: :functions
 import List, only: :macros
 ```
 
+It is also worth noting that import is lexical in scope.
```
Comment (doomspork) [inline lessons/basics/modules.md]: "that `import/2`"

File: lessons/basics/modules.md [hunk 2]
```diff
@@ -194,9 +202,28 @@ import List, only: :functions
...
+
+```elixir
+iex> defmodule MyModule do
+...>   def func_a do
+...>     import List, only: [duplicate: 2]
```
Comment (doomspork) [inline lessons/basics/modules.md]: I have not seen an `import/2` used within a function in any Elixir code, is this something we want to encourage?

File: lessons/basics/modules.md [hunk 3]
```diff
@@ -225,15 +254,58 @@ defmodule UseImportRequire.UseMe do
 end
 ```
 
-and we add this line to UseImportRequire:
+and we add this line to use UseImportRequire:
```
Comment (doomspork) [inline lessons/basics/modules.md]: "use `UseImportRequire`"

File: lessons/basics/modules.md [hunk 4]
```diff
@@ -225,15 +254,58 @@ defmodule UseImportRequire.UseMe do
...
-Using UseImportRequire.UseMe defines a use_test/0 function through invocation of the `__using__/1` macro.
+Using `UseImportRequire.UseMe` defines a `use_test/0` function through invocation of the `__using__/1` macro.
+
+```elixir
+iex> UseImportRequire.use_test
+use_test
+:ok
+```
+
+We can also override a given function that is given to use with `__using__/1`.
```
Comment (doomspork) [inline lessons/basics/modules.md]: I'm not sure about this, for one there are too many "givens" but this seems kind of out-of-place, there's no lead-in or reason.  Maybe we can reword it or maybe it should come at the end of the section on `use`?

How about something like:

"Sometimes it may be necessary to allow certain functions within our macro to be overridden, to this we'll need to use `defoverridable`.

Let's look at how we can leverage this:"

File: lessons/basics/modules.md [hunk 5]
```diff
@@ -225,15 +254,58 @@ defmodule UseImportRequire.UseMe do
...
+
+```elixir
+# lib/use_import_require/use_me.ex
+defmodule UseImportRequire.UseMe do
+  defmacro __using__(_) do
+    quote do
+      def use_test do
+        IO.puts "use_test"
+      end
+      defoverridable [use_test: 0]
```
Comment (doomspork) [inline lessons/basics/modules.md]: Does this need to come after the definition?  If so, we should mention that.

File: lessons/basics/modules.md [hunk 6]
```diff
@@ -225,15 +254,58 @@ defmodule UseImportRequire.UseMe do
...
+    end
+  end
+end
+```
+
+and now we can override it in IEx if we put it in another module:
```
Comment (doomspork) [inline lessons/basics/modules.md]: "Now let's override it in another module using IEx:"

File: lessons/basics/modules.md (hunk 1 above)
Comment (ybur-yug) [inline lessons/basics/modules.md]: Its straight out of the documentation:

```
iex(1)> h import

                         defmacro import(module, opts)

Imports functions and macros from other modules.

import/2 allows one to easily access functions or macros from others modules
without using the qualified name.
```

That said, if you dont like it I can see why.

File: lessons/basics/modules.md (hunk 5 above)
Comment (ybur-yug) [inline lessons/basics/modules.md]: I dont know but it has in the examples I saw from the docs. Will investigate.

File: lessons/basics/modules.md (hunk 4 above)
Comment (ybur-yug) [inline lessons/basics/modules.md]: Will work on something better than this now.

File: lessons/basics/modules.md (hunk 1 above)
Comment (ybur-yug) [inline lessons/basics/modules.md]: Misunderstood original comments concerns. I'm for backticks around them, I think, but dont believe we need arity (for things like `use` `require` `import` `__using__` etc.)

Comment (doomspork) [inline lessons/basics/modules.md]: Works for me.

File: lessons/basics/modules.md [hunk 7]
```diff
@@ -249,8 +288,44 @@ defmacro __using__(_) do
 end
 ```
 
-The `Ecto.Migration.__using__/1` macro includes an import call so that when you `use Ecto.Migration` you also `import Ecto.Migration`. It also sets up a module property which we will assume controls Ecto’s behavior.
+The `Ecto.Migration.__using__/1` macro includes an import call so that when you `use Ecto.Migration` you also `import Ecto.Migration`.
+It also sets up a module property which we will assume controls Ecto’s behavior.
+
+We can also override a given function that is given to use with `__using__/1`.
+This is done by using `defoverridable`.
+Lets implement this:
```
Comment (nscyclone) [inline lessons/basics/modules.md]: Shouldn't `Lets` be `Let's` in this case?

File: lessons/basics/modules.md [hunk 8]
```diff
@@ -249,8 +288,44 @@ defmacro __using__(_) do
...
+...> end
+{:module, Baz,
+ <<. . .>>, {:bar, 0}}
+iex> Baz.bar
+4
+```
+
+As you can see, we are able to override the default method `bar` from `use`, and even can call `super` to get the original functionality.
+It is worth noting that `super` must be called with the same number of arguments as the parent method being overridden.
+It also must be placed _after_ the functions being overridden.
```
Comment (nscyclone) [inline lessons/basics/modules.md]: I suppose we could also mention that `defoverridable` should not be used for private functions because this is deprecated since [1.4](https://github.com/elixir-lang/elixir/releases/tag/v1.4.0).

File: lessons/basics/modules.md [hunk 9]
```diff
@@ -249,8 +288,44 @@ defmacro __using__(_) do
...
 
-To recap: the use macro simply invokes the `__using__/1` macro of the specified module. To really understand what that does you need to read the `__using__/1` macro.
+### Finishing Up
+To recap: the `use` macro simply invokes the `__using__/1` macro of the specified module.
+To really understand what that does you need to read the `__using__/1` macro.
 
 **Note**: `quote`, `alias`, `use`, `require` are a macro used when we work with [metaprogramming](../../advanced/metaprogramming).
```
Comment (nscyclone) [inline lessons/basics/modules.md]: `are a macro` → `are macros`

File: lessons/basics/modules.md (hunk 8 above)
Comment (nscyclone) [inline lessons/basics/modules.md]: Probably we could also leave here a small section about `defdelegate` since we're adding `defoverridable`.

Comment (ybur-yug) [inline lessons/basics/modules.md]: I think `defdelegate` doesn't really fit. It prescribes behaviour to another module, but in this case we are trying to take behaviour already happening and extend/nullify/change it. I agree with the note on private functions. Thoughts @doomspork?

**Comment by ybur-yug** [comment]
Will make the grammar changes later today @nscyclone and we can all talk about the `defdelegate` bit too and decide what works best.

File: lessons/basics/modules.md (hunk 8 above)
Comment (doomspork) [inline lessons/basics/modules.md]: This is a good question.  I'm not sure `defdelegate` makes sense in _this_ portion of the lesson but if we're expanding the Module content I don't see why we shouldn't find a way to mention `defdelegate`.

File: lessons/basics/modules.md [hunk 10]
```diff
@@ -249,8 +288,44 @@ defmacro __using__(_) do
...
+...>       defoverridable [bar: 0]
+...>     end
+...>   end
+...> end
+{:module, Foo,
+ <<. . .>>, {:__using__, 1}}
+iex> defmodule Baz do
+...>   use Foo
+...>   def bar do
+...>     super + 2
```
Comment (nscyclone) [inline lessons/basics/modules.md]: I think `super` should be `super()` here.

Comment (ybur-yug) [inline lessons/basics/modules.md]: Youre right, in 1.4 it will warn w/o it.

Comment (ybur-yug) [inline lessons/basics/modules.md]: I'm gonna go through and edit RE suggestions and such later tonight or tomorrow at the airport and push everything up.

**Comment by ybur-yug** [comment]
@doomspork @nscyclone I believe I addressed all the points you both brought up, please take a look and see what you think now that things are updated when you've a chance.

File: lessons/basics/modules.md (hunk 2 above)
Comment (ybur-yug) [inline lessons/basics/modules.md]: I decided to leave this because it is specific to the behaviour of `import` itself, and worth noting. But now that I think more, its also true that `require` and `import` are also lexical. Maybe I should edit to just make one blanket statement about all of them, and not encourage it but just make it known.

Thoughts?

Comment (ybur-yug) [inline lessons/basics/modules.md]: CC @nscyclone @brain-geek or anyone else who has an opinion

Comment (doomspork) [inline lessons/basics/modules.md]: Let's remove this.  I don't like it and I don't want to encourage this nor confuse folks.

File: lessons/basics/modules.md [hunk 11]
```diff
@@ -249,8 +288,66 @@ defmacro __using__(_) do
...
+...>     super() + 2
+...>   end
+...> end
+{:module, Baz,
+ <<. . .>>, {:bar, 0}}
+iex> Baz.bar
+4
+```
+
+As you can see, we are able to override the default method `bar` from `use`, and even can call `super` to get the original functionality.
```
Comment (doomspork) [inline lessons/basics/modules.md]: "As you can see are able to override the default method `bar/0`"

File: lessons/basics/modules.md [hunk 12]
```diff
@@ -249,8 +288,66 @@ defmacro __using__(_) do
...
+It is worth noting that `super` must be called with the same number of arguments as the parent method being overridden.
+It also must be placed _after_ the functions being overridden.
+
+### Delegation
+It is possible to define a function that delegates to another module as well.
```
Comment (doomspork) [inline lessons/basics/modules.md]: Let's drop the "as well".

File: lessons/basics/modules.md [hunk 13]
```diff
@@ -249,8 +288,66 @@ defmacro __using__(_) do
...
+If you use `defdelegate` instead of `def`, a given function will be public and invoked from outside the module it is defined in.
```
Comment (doomspork) [inline lessons/basics/modules.md]: The "invoked from outside the module" portion of this is a little confusing to me.  How about something like:

"Using `defdelegate` we can define a public function and delegate it to another module's function, let's see it in action:"

File: lessons/basics/modules.md [hunk 14]
```diff
@@ -249,8 +288,66 @@ defmacro __using__(_) do
...
+iex> defmodule SomeOtherModule do
+...>   defdelegate world, to: MyModule, as: :hello
+...> end
+{:module, SomeOtherModule,
+ <<70, ... >>, [world: 0]}
+iex> IO.puts SomeOtherModule.world
+hello from my module
+```
+
+As you can see, the code is called from `MyModule` when called through `SomeOtherModule`.
```
Comment (doomspork) [inline lessons/basics/modules.md]: "As you can see our calls to `SomeOtherModule.world/0` are delegated to `MyModule.hello/0`."

File: lessons/basics/modules.md (hunk 2 above)
Comment (brain-geek) [inline lessons/basics/modules.md]: I've seen it used in some cases when otherwise you'll have namespace conflicts.

File: lessons/basics/modules.md [hunk 15]
```diff
@@ -6,13 +6,15 @@ order: 8
 lang: en
 ---
 
-We know from experience it's unruly to have all of our functions in the same file and scope.  In this lesson we're going to cover how to group functions and define a specialized map known as a struct in order to organize our code more efficiently.
+We know from experience it's unruly to have all of our functions in the same file and scope.
```
Comment (doomspork) [inline lessons/basics/modules.md]: Why are we putting every sentence on a new line?  This creates a huge amount of unnecessary changes on each PR.  If there isn't a good reason for this let's stop.

File: lessons/basics/modules.md [hunk 16]
```diff
@@ -249,8 +288,65 @@ defmacro __using__(_) do
...
-The `Ecto.Migration.__using__/1` macro includes an import call so that when you `use Ecto.Migration` you also `import Ecto.Migration`. It also sets up a module property which we will assume controls Ecto’s behavior.
+The `Ecto.Migration.__using__/1` macro includes an import call so that when you `use Ecto.Migration` you also `import Ecto.Migration`.
+It also sets up a module property which we will assume controls Ecto’s behavior.
+
+We can also override a given function that is given to use with `__using__/1`.
+This is done by using `defoverridable`.
+Let's implement this:
+
+```elixir
+iex> defmodule Foo do
```
Comment (doomspork) [inline lessons/basics/modules.md]: Traditionally we've not put module code within `iex` because it becomes more difficult to read.

Let's try to keep the lesson styles the same

File: lessons/basics/modules.md [hunk 17]
```diff
@@ -251,6 +273,56 @@ end
...
+...> end
+{:module, SomeOtherModule,
+ <<70, ... >>, [world: 0]}
+iex> IO.puts SomeOtherModule.world
+hello from my module
+```
+
+As you can see our calls to `SomeOtherModule.world/0` are delegated to `MyModule.hello/0`.
+
+### Finishing Up
```
Comment (doomspork) [inline lessons/basics/modules.md]: Let's remove this section header and get this merged 👍

**Comment by ybur-yug** [comment]
@doomspork last feedback incorporated + pushed

**Comment by doomspork** [comment]
I'm going to close this for now since it needs to be rebased and have feedback incorporated.  @ybur-yug let's sync up in Slack about these updates and some other changes for 1.5 😀
